x/vote/types/vote_router.go

```diff
@@ -74,12 +74,29 @@ type handlerWrapper struct {
 }
 
 func (w handlerWrapper) HandleResult(ctx sdk.Context, result codec.ProtoMarshaler) error {
-	cachedCtx, writeCache := ctx.CacheContext()
-	if err := w.VoteHandler.HandleResult(cachedCtx, result); err != nil {
-		return err
-	}
+	return cache(w.VoteHandler.HandleResult)(ctx, result)
+}
 
-	writeCache()
-	ctx.EventManager().EmitEvents(cachedCtx.EventManager().Events())
-	return nil
+func (w handlerWrapper) HandleFailedPoll(ctx sdk.Context, poll exported.Poll) error {
+	return cache(w.VoteHandler.HandleFailedPoll)(ctx, poll)
+}
+
+func (w handlerWrapper) HandleCompletedPoll(ctx sdk.Context, poll exported.Poll) error {
+	return cache(w.VoteHandler.HandleCompletedPoll)(ctx, poll)
+}
+
+func (w handlerWrapper) HandleExpiredPoll(ctx sdk.Context, poll exported.Poll) error {
+	return cache(w.VoteHandler.HandleExpiredPoll)(ctx, poll)
+}
+
+func cache[T any](f func(ctx sdk.Context, x T) error) func(ctx sdk.Context, x T) error {
+	return func(ctx sdk.Context, x T) error {
+		cachedCtx, writeCache := ctx.CacheContext()
+		if err := f(cachedCtx, x); err != nil {
```

Comment:
Maybe use `RunCached` instead? It'll also enable protection against panics.

Reply:
which is precisely what I don't want, that's why I didn't use it

Reply:
this is not the place to suppress panics

Reply:
Hmm, but then if there's a panic during vote handling the validator will be the one that gets penalized for it though, since the panic will propagate to the VoteRequest. Instead of outright suppressing, we can return an error instead. Note that VoteRequest does eventually suppress the error (but not panics).

Reply:
after discussion we agreed to do a clean refactoring in the next release instead of a quick band-aid now since this is very low impact